src/patch.rs

```diff
@@ -0,0 +1,203 @@
+use libc::{c_char, c_int, c_void};
+
+use {raw, Blob, Buf, Diff, DiffDelta, DiffHunk, DiffLine, DiffOptions, Error, IntoCString};
+use diff::{LineCb, print_cb};
+use util::{Binding, into_opt_c_string};
+
+/// A structure representing the text changes in a single diff delta.
+///
+/// This is an opaque structure.
+pub struct Patch {
+    raw: *mut raw::git_patch,
+}
+
+unsafe impl Send for Patch {}
+
+impl Binding for Patch {
+    type Raw = *mut raw::git_patch;
+    unsafe fn from_raw(raw: Self::Raw) -> Patch {
+        Patch { raw: raw }
+    }
+    fn raw(&self) -> Self::Raw { self.raw }
+}
+
+impl Drop for Patch {
+    fn drop(&mut self) {
+        unsafe { raw::git_patch_free(self.raw) }
+    }
+}
+
+impl Patch {
+    /// Return a Patch for one file in a Diff.
+    ///
+    /// Returns Ok(None) for an unchanged or binary file.
+    pub fn from_diff(diff: &Diff, idx: usize) -> Result<Option<Patch>, Error> {
+        let mut ret = 0 as *mut raw::git_patch;
+        unsafe {
+            try_call!(raw::git_patch_from_diff(&mut ret, diff.raw(), idx));
+            Ok(Binding::from_raw_opt(ret))
+        }
+    }
+
+    /// Generate a Patch by diffing two blobs.
+    pub fn from_blobs<S1, S2>(old_blob: &Blob,
+                              old_path: Option<S1>,
```

Comment:
Perhaps these could be `Option<&Path>` to help with type inference when `None` is passed?